File: lang/c/src/vpkppc/format/FPX.cpp
#include <vpkppc/format/FPX.h>

#include <sourceppc/Helpers.h>

using namespace sourceppc;
using namespace vpkpp;

const uint32_t VPKPP_FPX_SIGNATURE = FPX_SIGNATURE;
const char* VPKPP_FPX_DIR_SUFFIX   = FPX_DIR_SUFFIX.data();
const char* VPKPP_FPX_EXTENSION    = FPX_EXTENSION.data();

SOURCEPP_API vpkpp_pack_file_handle_t vpkpp_fpx_create(const char* path) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = FPX::create(path);
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}

SOURCEPP_API vpkpp_pack_file_handle_t vpkpp_fpx_open(const char* path, vpkpp_entry_callback_t callback) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = FPX::open(path, callback ? [callback](const std::string& entryPath, const Entry& entry) {
		callback(entryPath.c_str(), const_cast<Entry*>(&entry));
	} : static_cast<PackFile::EntryCallback>(nullptr));
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}

File: lang/csharp/src/vpkpp/Format/FPX.cs
using System;

namespace sourcepp.vpkpp.Format;

using OpenPropertyRequest = Func<PackFile, OpenProperty, byte[]>;

using EntryCallback = Action<string, Entry>;

public class FPX : PackFile
{
	protected FPX(nint handle, bool managed = true) : base(handle, managed)
	{
	}

	public static FPX? Create(string path)
	{
		var handle = DLL.vpkpp_fpx_create(path);
		return handle == nint.Zero ? null : new FPX(handle);
	}

	public new static FPX? Open(string path, EntryCallback? callback = null, OpenPropertyRequest? _ = null)
	{
		var handle = DLL.vpkpp_fpx_open(path, callback is not null ? (entryPath, entry) =>
		{
			callback(entryPath, new Entry(entry, false));
		} : null);
		return handle == nint.Zero ? null : new FPX(handle);
	}
}
